2단계 - 문자열 덧셈 계산기

h2jinee opened the pull request:
## 📌 Summary
- 1단계 리팩토링을 진행했습니다.
- 2단계 학습 테스트 실습을 완료했습니다. (회사 일이 바빠서 늦었습니다 🥹)
    - TDD 사이클(Test -> Implement -> Refactor)을 따르며 바텀업 방식으로 구현을 진행했습니다.
- 제가 요구사항을 제대로 이해하지 못해서 `input => sum` 템플릿으로 만들라고 이해를 해서 처음에 그 방식대로 구현을 했다가 뒤늦게 아닌 것을 알게 되어서 수정했습니다 😂

## 💬 Review Points
- **관심사 분리:**
    - `Calculator` (계산 로직)와 `Delimiter` (구분자 처리 로직) 클래스로 역할을 분리하여 구현했습니다.
    - 객체지향적 설계가 아직 미숙해서 `Delimiter` 클래스는 구현 과정에서 AI의 도움을 조금 받았습니다. 😅 
    - 객체지향적 관점에서 아쉬운 부분이 있는지 리뷰를 받고 싶습니다.
- **테스트:**
    - 매번 `Calculator` 객체를 생성하는 것이 비효율적이라 생각되어 `@BeforeEach`를 사용했습니다.
    - 합계 테스트를 위해  `@MethodSource`를 사용해보았습니다.
- **기능:**
    - `//`와 `\n` 사이의 문자열을 커스텀 구분자로 처리하는 기능을 구현했습니다.
    - `Delimiter` 클래스 내에 기본 구분자(`,` 와`:`)를 상수로 선언하여 관리하도록 했습니다.

### **🤔 고민한 점 및 질문**
- **Arrange-Act-Assert 주석 피드백 관련**
    - 개인적으로 3라인 이상이 넘어갈 경우 주석을 작성하는 것이 더 눈에 잘 들어오는 것 같습니다. 2라인까지는 재성님께서 피드백 주셨던 것처럼 제거하는 게 더 깔끔하고, 이해하기에도 어렵지 않다는 생각이 들었습니다!

- **테스트 검증 범위 관련**
    - 초기에 요구사항을 `input => sum` 템플릿으로 만들라는 것으로 잘못 이해했습니다.
    - 만약 요구사항이 `input => sum` 템플릿으로 만드는 것이었다면, **빈 문자열 또는 null 값을 입력할 경우 0을 반환한다**을 테스트할 때 최종 반환 값(`input => sum`)의 형태로 검증하는 것이 맞을지, 아니면 내부 헬퍼 메서드 `parseOrZero`의 동작만 테스트 하는 것이 맞는지 궁금합니다.

- **테스트 케이스 중복 관련**
    - **숫자 두개를 컴마(,) 또는 콜론(:) 구분자로 입력할 경우 두 숫자의 합을 반환한다** 테스트와 **구분자는 컴마(,)와 콜론(:)을 사용할 수 있다** 이 두 테스트가 사실상 동일한 내용을 검증하는 중복 테스트이지 않나? 싶은 생각이 들었습니다. 재성님께서는 어떻게 생각하시는지 궁금합니다.
    
- **유효성 검증 로직 관련**
    - `calculate` 메서드 내에 `parseOrZero`, `parseIfSingleDigit` 같은 유효성 검증 로직을 모두 포함하면 메서드가 너무 길어질 것 같아 별도 메서드로 분리했습니다.
    - 현재처럼 분리하는 것이 응집도나 가독성 면에서 더 나은 선택일까요?
    - `parseOrZero`, `parseIfSingleDigit` 이 두 메서드의 경우에도 하나로 합치는 것이 더 좋은 방법일까요?


## 📎Reference
- [AssertJ Exception Assertions](https://www.baeldung.com/assertj-exception-assertion)

h2jinee committed:
refactor: 1단계 - 학습 테스트 피드백 반영

h2jinee committed:
test: 2단계 - 문자열 덧셈 계산기

h2jinee committed:
refactor: formatResult 제거

h2jinee committed:
refactor: 불필요한 주석 제거

h2jinee committed:
refactor: 변수명 수정

javajigi commented on src/main/java/step2/Calculator.java:
```diff
@@ -0,0 +1,53 @@
+package step2;
+
+public class Calculator {
+	private final Delimiter delimiter;
```
Delimiter 추가는 신선한 접근 방식이네요. 👍

javajigi commented on src/main/java/step2/Calculator.java:
```diff
@@ -0,0 +1,53 @@
+		this.delimiter = delimiter;
+	}
+
+	public int calculate(String input) {
```
```suggestion
		if (isBlank(input)) {
			return 0;
		}
```

parseOrZero을 추가하기보다 위와 같이 calculate() 내에서 처리하는 것은 어떨까?

javajigi commented on src/main/java/step2/Calculator.java:
```diff
@@ -0,0 +1,53 @@
+		return Integer.parseInt(input);
+	}
+
+	public int parseIfSingleDigit(String input) {
```
```suggestion
	String[] values = "1".split( delimiter);
```
숫자가 하나인 경우를 split할 경우 위와 같아 굳이 별도의 메서드로 분리하지 않아도 됨

javajigi commented on src/main/java/step2/Calculator.java:
```diff
@@ -0,0 +1,53 @@
+
+			if (number < 0) {
+				throw new IllegalArgumentException("음수는 입력할 수 없습니다.");
+			}
```
indent(들여쓰기)를 줄이기 위해 메서드를 분리해 보는 것은 어떨까?

javajigi commented on src/main/java/step2/Calculator.java:
```diff
@@ -0,0 +1,53 @@
+		if (!input.matches(regex)) {
+			throw new IllegalArgumentException("허용되지 않은 구분자가 포함되어 있습니다.");
+		}
+	}
```
지금의 경우 각 단위 테스트 코드에서 분산되어 각각의 메서드를 테스트하고 있는데요.
어떤 delimiter를 생성할 것인지 결정하고, null or 빈 문자열 체크, 합을 구하는 모든 로직이 calculate() 메서드 내에서 구현해야해요.
즉, calculate() 메서드만 public이고, 나머지 분리한 메서드는 private으로 구현해 보면 테스트를 어떻게 해야할 것인지 힌트를 얻을 수 있을 겁니다.

h2jinee replied:
재성님 이 코멘트를 보니 감이 잡혔습니다 감사합니다!!

javajigi commented on src/test/java/step2/CalculatorTest.java:
```diff
@@ -0,0 +1,114 @@
+package step2;
```
step2보다는 caclulator와 같은 기능 이름이면 더 좋지 않을까요?

javajigi reviewed (changes requested):
바쁜 와중에도 2단계 구현 👍 
전체적으로 메서드 분리 잘 했네요.
그런데 이 미션 진행할 때, TDD 접근할 때 희진님처럼 접근하는 경우가 있는데요.
피드백 남겼으니 calculate를 통해 모든 테스트가 가능하도록 도전해 보세요.
대부분의 기능이 이미 구현되어 있으니 쉽게 리팩터링할 수 있을 겁니다.

Q1. Arrange-Act-Assert 주석 피드백 관련
제 생각을 피드백으로 남긴 것이니까요. 희진님이 판단해 보시고 선호하는 방향으로 구현하면 됩니다.

Q2. 테스트 검증 범위 관련
접근 방식을 다르게 생각하고 있어 나왔던 질문으로 보여요.
피드백에 남긴 내용 반영하다보면 쉽게 해결될 겁니다.

Q3. 테스트 케이스 중복 관련
맞습니다. 둘이 하나의 test case로 구현해도 되고, 분리해서 구현해도 된다 생각해요.

Q4. 유효성 검증 로직 관련
이 또한 접근 방식을 다르게 생각하고 있어 나왔던 질문으로 보여요.
피드백에 남긴 내용 반영하다보면 고민이 풀리지 않을까 생각되네요.

h2jinee committed:
refactor: 피드백 반영 - calculate() 단일 public 메서드로 캡슐화 및 테스트 코드 수정

- calculate() 메서드만 public으로 노출, 나머지 메서드는 private으로 캡슐화
- 개별 메서드 테스트를 제거하고 calculate() 메서드 통합 테스트로 변경
- step2 패키지를 calculator로 이름 변경

javajigi commented on src/main/java/calculator/Calculator.java:
```diff
@@ -0,0 +1,57 @@
+		this.delimiter = delimiter;
+	}
+
+	public int calculate(String input) {
```
이 메서드도 추가 리팩터링이 가능할 것 같아요.
3단계 진행하는 틈틈이 지난 주 첫번째 라이브 강의 영상 한번 봐도 도움이 될 것 같아요.

javajigi commented on src/main/java/calculator/Calculator.java:
```diff
@@ -0,0 +1,57 @@
+		}
+	}
+
+	private void validatePositiveNumber(String token) {
```
👍

javajigi reviewed (approved):
리팩터링 방향 잘 잡고 진행했네요. 👍 
input 문자열에 따라 Delimiter를 결정하는 부분도 calculate() 내에 구현하는 것이 맞는 방향입니다.

2단계는 이 정도로 진행하고 실제 현실의 요구사항과 비슷한 3단계 도전해 보시죠.
3단계는 단위 테스트, TDD도 중요하지만 그 보다 간단한 객체 설계라도 도전해 보면 좋겠습니다.
